rocAL/include/pipeline/node.h

```diff
@@ -73,3 +75,62 @@ class Node {
         (this->_args.push_back(Argument(arg_names[Indices], std::forward<Args>(args))), ...);
     }
 };
+
+class NodeFactory {
+public:
+    using LoaderCreator = std::function<std::shared_ptr<Node>(Tensor*, void*)>;
+    using AugmentationCreator = std::function<std::shared_ptr<Node>(const std::vector<Tensor *>&, const std::vector<Tensor *>&)>;
+
+    static NodeFactory& instance() {
+        static NodeFactory factory;
+        return factory;
+    }
+
+    void register_loader_node(const std::string& name, LoaderCreator creator) {
+        _loader_node_registry[name] = std::move(creator);
+    }
+
+    void register_node(const std::string& name, AugmentationCreator creator) {
+        _node_registry[name] = std::move(creator);
+    }
+
+    std::shared_ptr<Node> create_loader_node(const std::string& name, Tensor* output_tensor, void *dev_resource) const {
+        auto it = _loader_node_registry.find(name);
+        if (it != _loader_node_registry.end()) {
+            return it->second(output_tensor, dev_resource);
+        } else {
+            THROW("The given node not found in the registry" + name)
+        }
+    }
+
+    std::shared_ptr<Node> create_node(const std::string& name, const std::vector<Tensor *>& inputs, const std::vector<Tensor *>& outputs) const {
+        auto it = _node_registry.find(name);
+        if (it != _node_registry.end()) {
+            return it->second(inputs, outputs);
+        } else {
+            THROW("The given node not found in the registry" + name)
+        }
+    }
+
+private:
```

Comment:
Non-blocking: `NodeFactory` relies on `std::map`/`std::function`, but this header does not include `<map>` or `<functional>` directly. Today it works only because the transitive includes from other headers happen to pull them in. Could we add the explicit includes so the header stays self-contained?